lib/ffi/src/lib.rs

```diff
@@ -0,0 +1,203 @@
+//! Rusty Vangers FFI bindings.
+//! Matches "lib/renderer/src/renderer/scene/rust/vange_rs.h"
+//! See https://github.com/KranX/Vangers/pull/517
+
+use futures::executor::LocalPool;
+use std::ptr;
+
+#[repr(C)]
+#[derive(Default)]
+pub struct Vector3 {
+    x: f32,
+    y: f32,
+    z: f32,
+}
+
+#[repr(C)]
+#[derive(Default)]
+pub struct Quaternion {
+    x: f32,
+    y: f32,
+    z: f32,
+    w: f32,
+}
+
+#[repr(C)]
+#[derive(Default)]
+pub struct Transform {
+    position: Vector3,
+    rotation: Quaternion,
+}
+
+#[repr(C)]
+pub struct Rect {
+    x: i32,
+    y: i32,
+    width: i32,
+    height: i32,
+}
+
+#[repr(C)]
+#[derive(Default)]
+pub struct CameraDescription {
+    fov: f32,
+    aspect: f32,
+    near: f32,
+    far: f32,
+}
+
+#[repr(C)]
+#[derive(Clone, Copy)]
+pub struct MapDescription {
+    width: i32,
+    height: i32,
+    lines: *const *const u8,
+    material_begin_offsets: *const u8,
+    material_end_offsets: *const u8,
+    material_count: i32,
+}
+
+#[derive(Default)]
+struct Camera {
+    desc: CameraDescription,
+    transform: Transform,
+}
+
+struct LevelContext {
+    desc: MapDescription,
+    render: vangers::render::Render,
+    level: vangers::level::Level,
+}
+
+pub struct Context {
+    level: Option<LevelContext>,
+    render_config: vangers::config::settings::Render,
+    color_format: wgpu::TextureFormat,
+    queue: wgpu::Queue,
+    device: wgpu::Device,
+    downlevel_caps: wgpu::DownlevelCapabilities,
+    _instance: wgpu::Instance,
+    camera: Camera,
+}
+
+#[no_mangle]
+pub extern "C" fn rv_init() -> Option<ptr::NonNull<Context>> {
+    use vangers::config::settings as st;
+
+    let mut task_pool = LocalPool::new();
```

Comment:
Is there any advantage to use task_pool instead of async/await? LocalPool didn't work in wasm target, but await/async works. So it's preffered to use await/async if possible. Like:
```rust
adapter.request_device().await
```

Reply:
We can't do `await` inside a synchronous function. And function is synchronous because it's `pub extern "C"`, so I don't think there is much of a choice here.